File: 0501-Find_Mode_in_Binary_Search_Tree/Find_Mode_in_Binary_Search_Tree.go
package findmodeinbinarysearchtree

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var maxOccurrence int

func findMode(root *TreeNode) []int {
	var dp = make(map[int]int)
	maxOccurrence = 0

	dfs(root, dp)

	result := make([]int, 0)
	for key, value := range dp {
		if value == maxOccurrence {
			result = append(result, key)
		}
	}

	return result
}

func dfs(root *TreeNode, dp map[int]int) {
	if root != nil {
		curr := dp[root.Val]
		curr++
		dp[root.Val] = curr

		maxOccurrence = max(maxOccurrence, curr)

		dfs(root.Left, dp)
		dfs(root.Right, dp)
	}

}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 0501-Find_Mode_in_Binary_Search_Tree/Find_Mode_in_Binary_Search_Tree.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
 
class Solution {
    HashMap<Integer,Integer> hm = new HashMap<>();
    int max = 0;
    public int[] findMode(TreeNode root) {
        getOccurrence(root);
        List<Integer> ans = new ArrayList<Integer>();
        
        for(Integer i: hm.keySet()){
            if(hm.get(i) == max){
                ans.add(i);
            }
        }
        
        int[] ansArr = new int[ans.size()];
        
        for(int i = 0; i < ans.size(); i++){
            ansArr[i] = ans.get(i);
        }
        return ansArr;
    }
    
    private void getOccurrence(TreeNode root){
        if(root == null) {
            return;
        }
        
        int curr = hm.getOrDefault(root.val, 0) + 1;
        hm.put(root.val, curr);
        max = Math.max(max, curr);
        
        getOccurrence(root.left);
        getOccurrence(root.right);
    }
}